cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/WrappedSQLEngineCollection.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.spark.batch;
+
+import io.cdap.cdap.etl.api.batch.SparkCompute;
+import io.cdap.cdap.etl.api.batch.SparkSink;
+import io.cdap.cdap.etl.api.streaming.Windower;
+import io.cdap.cdap.etl.common.PhaseSpec;
+import io.cdap.cdap.etl.common.RecordInfo;
+import io.cdap.cdap.etl.common.StageStatisticsCollector;
+import io.cdap.cdap.etl.proto.v2.spec.StageSpec;
+import io.cdap.cdap.etl.spark.SparkCollection;
+import io.cdap.cdap.etl.spark.SparkPairCollection;
+import io.cdap.cdap.etl.spark.join.JoinExpressionRequest;
+import io.cdap.cdap.etl.spark.join.JoinRequest;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.PairFlatMapFunction;
+
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * SQLEngineBackedCollection that wraps another SQLEngineBackedCollection in order to delay the execution of a
+ * mapping function.
+ *
+ * This is currently used to prevent SQL Engine pull operations uniless absolutely needed.
+ * @param <T> Type of the wrapped collection records.
+ * @param <U> Type of the output collection records.
+ */
+public class WrappedSQLEngineCollection<T, U> implements SQLBackedCollection<U> {
+  private final java.util.function.Function<SparkCollection<T>, SparkCollection<U>> mapper;
+  private final SQLBackedCollection<T> wrapped;
+  private SparkCollection<U> unwrapped = null;
+
+  public WrappedSQLEngineCollection(SQLBackedCollection<T> wrapped,
+                                    java.util.function.Function<SparkCollection<T>, SparkCollection<U>> mapper) {
```

Comment:
Any reason you don't want to go with just Supplier?

Reply:
No particular reason, I'll refactor to use supplier and see how this looks.

Reply:
Yeah, would be good. Function is fine, but I don't like two Function classes in one file.

Reply:
I worked on the refactor and I see an issue at this point:

Using a Function, I can enforce the rule that the wrapped collection is an instance of `SQLBackedCollection`.

If I use a supplier, the supplier method will need to return an instance of the base interface, which is `SparkCollection`.

This has the undesired side effect that I cannot really ensure that the collection I'm wrapping is really a SQL backed collection. It's not a huge issue if the users of this class are aware of the reason for it's existence, but it's not something I can set in the constructor "contract".

The code would look something like:

```
public class WrappedSQLEngineCollection<T, U> implements SQLBackedCollection<U> {
  private final Supplier<SparkCollection<U>> mapper;
  private SparkCollection<U> unwrapped = null;

  public WrappedSQLEngineCollection(Supplier<SparkCollection<U>> mapper) {
    this.mapper = mapper;
  }

  private SparkCollection<U> unwrap() {
    if (unwrapped == null) {
      unwrapped = mapper.get();
    }

    return unwrapped;
  }

  @Override
  public <C> C getUnderlying() {
    return unwrap().getUnderlying();
  }
  ...
```
And it would be instantiated by using something like:
```
    if (collection instanceof SQLBackedCollection) {
      return new WrappedSQLEngineCollection<>(() -> collection.map(new RecordInfoWrapper<>(stageName)));
    }
```

@tivv let me know your thoughts.

Reply:
Using marker interfaces is often tricky with wrapping and delegation. Have you considered adding a "hasCapability(CapabilityEnum e)" to the SparkCollection or something along the lines? This approach is often much easier to support for delegates.

Reply:
Hmm, good idea. I had not considered this.

Reply:
@albertshau any thoughts on @tivv suggestion? I might implement this change in a later PR. For now, I'd just like to  merge this and continue focusing on BQ Pushdown